File: HW8/hw8.py
from collections import Counter

class hw8():
    def longestPalindrome(self, s: str) -> int:
        char_counts = Counter(s)
        palindrome_length = 0
        odd_found = False
        
        for count in char_counts.values():
            if count % 2 == 0:
                palindrome_length += count
            else:
                palindrome_length += count - 1  # use all except one to make it even
                odd_found = True  # at least one odd count character exists
        
        # if there odd count character, one can be the center
        if odd_found:
            palindrome_length += 1
        
        return palindrome_length
    
# test = hw8()
# s = "abccccdd"
# t ="speediskey"
# print(test.longestPalindrome(t))

File: HW8/hw8.java
import java.util.*;

public class hw8 {
    public int longestPalindrome(String s) {
        Map<Character, Integer> countMap = new HashMap<>();
        
        // Count the frequency of each character
        for (char c : s.toCharArray()) {
            countMap.put(c, countMap.getOrDefault(c, 0) + 1);
        }
        
        int maxLength = 0;
        boolean hasOdd = false;
        
        // Calculate the maximum length of the palindrome
        for (int count : countMap.values()) {
            if (count % 2 == 0) {
                maxLength += count;
            } else {
                maxLength += count - 1; // Add the largest even number less than count
                hasOdd = true; // There's at least one character with an odd count
            }
        }
        
        // If there's at least one character with an odd count, we can place exactly one character in the middle
        if (hasOdd) {
            maxLength += 1;
        }
        
        return maxLength;
    }

    public static void main(String[] args){
        // hw8 test = new hw8();
        // String s = "abccccdd";
        // String t = "speediskey";
        // System.out.println(test.longestPalindrome(t));
    }
}
